File: app/javascript/embeds/tidal.js
import MediaSite from "media_site"

class Tidal extends MediaSite {
  constructor() {
    super(
      /https:\/\/tidal\.com\/(playlist\/([a-zA-Z0-9\-]+)|(browse\/)?(track|album)\/([0-9]+))(\?.*)?/,

      async (url) => {
        const match = url.match(this.regex)
        if (match) {
          if (match[2]) {
            // Playlist format: https://tidal.com/playlist/id
            const playlistId = match[2]
            return `https://embed.tidal.com/playlists/${playlistId}`
          } else if (match[4] && match[5]) {
            // Track/Album format: https://tidal.com/browse/track/id or https://tidal.com/track/id
            const type = match[4]
            const id = match[5]
            const pluralType = type === 'track' ? 'tracks' : 'albums'
            return `https://embed.tidal.com/${pluralType}/${id}`
          }
        }
        return null
      },

      (embedUrl) => {
        const iframe = document.createElement('iframe')
        iframe.src = embedUrl
        iframe.allow = "encrypted-media"
        iframe.sandbox = "allow-same-origin allow-scripts allow-forms allow-popups"
        iframe.title = "TIDAL Embed Player"
        iframe.loading = "lazy"
        iframe.style.display = "block"

        // Set height based on content type
        // Tidal embeds have internal 100vh sizing, so these need to be generous
        if (embedUrl.includes('/playlists/')) {
          iframe.height = "600"
        } else if (embedUrl.includes('/albums/')) {
          iframe.height = "450"
        } else {
          iframe.height = "120" // tracks - compact player
        }

        return iframe
      }
    )
  }
}

export default Tidal

File: app/models/media_embed/providers/tidal.rb
module MediaEmbed
  module Providers
    class Tidal < Base
      REGEX = %r{\Ahttps://tidal\.com/(?:playlist/([a-zA-Z0-9-]+)|(?:browse/)?(track|album)/([0-9]+))}i

      def render(url)
        match = url.match(REGEX)
        return unless match

        playlist_id, type, id = match.captures
        embed_url = if playlist_id
          "https://embed.tidal.com/playlists/#{playlist_id}"
        else
          "https://embed.tidal.com/#{type == "track" ? "tracks" : "albums"}/#{id}"
        end

        iframe(embed_url,
          height: tidal_height(embed_url),
          allow: "encrypted-media",
          sandbox: "allow-same-origin allow-scripts allow-forms allow-popups",
          title: "TIDAL Embed Player",
          style: style("display: block", "max-width: 660px", "margin-inline: auto"))
      end

      private

        def tidal_height(embed_url)
          return "600" if embed_url.include?("/playlists/")
          return "450" if embed_url.include?("/albums/")

          "120"
        end
    end
  end
end
